cloudinit/config/schema.py

```diff
@@ -96,21 +101,35 @@ def validate_cloudconfig_schema(config, schema, strict=False):
     # Allow for bytes to be presented as an acceptable valid value for string
     # type jsonschema attributes in cloud-init's schema.
     # This allows #cloud-config to provide valid yaml "content: !!binary | ..."
+
     if hasattr(Draft4Validator, 'TYPE_CHECKER'):  # jsonschema 3.0+
         type_checker = Draft4Validator.TYPE_CHECKER.redefine(
             'string', is_schema_byte_string)
         cloudinitValidator = extend(Draft4Validator, type_checker=type_checker)
     else:  # jsonschema 2.6 workaround
         types = Draft4Validator.DEFAULT_TYPES
-        # Allow bytes as well as string (and disable a spurious
-        # unsupported-assignment-operation pylint warning which appears because
-        # this code path isn't written against the latest jsonschema).
+        # Allow bytes as well as string (and disable a spurious unsupported
+        # assignment-operation pylint warning which appears because this
+        # code path isn't written against the latest jsonschema).
         types['string'] = (str, bytes)  # pylint: disable=E1137
         cloudinitValidator = create(
             meta_schema=Draft4Validator.META_SCHEMA,
             validators=Draft4Validator.VALIDATORS,
             version="draft4",
             default_types=types)
+
+    if strict_metaschema:
+        mymeta = cloudinitValidator.META_SCHEMA
+
+        # this disables bottom-level keys
+        mymeta['additionalProperties'] = False
+
+        # encoding the base level jsonschema definitions
+        # necessary since (since additionalProperties=False)
+        cloudinitValidator.check_schema(schema)
+
+        cloudinitValidator(schema, format_checker=FormatChecker())
+        return
```

Comment:
> @holmanb What do you feel is the use-case we are solving where we want the early exit and ignore of the schema validation below if strict_metaschema=True?

My bad, the intention is not clear here.

Originally this was a separate function in the unittest, but I shaved about 70LOC from the diff by making it a single `if` statement in `schema.py`. I'm realizing that perhaps refactoring out a separate function `get_schema_validator()` for use in `validate_cloudconfig_schema()` and the unittest would probably have been better than muddying the purpose of `validate_cloudconfig_schema()` by making it dual-purpose.

Basically the logic is currently:
```
if strict_metaschema:
    <validate the schema using the metaschema>
    return
<validate the config using the schema>
```

maybe clearer would have been:
```
if strict_metaschema:
    <validate schema using metaschema>
else:
    <validate config using schema>
```
The unittest passes an empty config when calling `validate_cloudconfig_schema()`, since the purpose of that tests is to validate the schemas, not to validate a config with the schema.

> And is it worth a unittest to show that error conditions in schema don't get raised with strict_metaschema=True. (such a unit test docstring would also be a place to document why we don't want to raise errors at runtime).

This sounds like a valuable addition. I will add that.

Reply:
@holmanb I'm defintiely all for early exits specifically for avoiding nested scopes too.  It feel clearer in my opinion as we can then avoid additional line-wraps due to overflow due to all the extra indents later in functions/methods. All good there.  I don't really mind the dual-purpose of the validate_cloudconfig_schema(), that seems to make sense to me as we are generally validating different scopes of the "schema".

My comment really was that don't see why the two options need to be mutually-exclusive. If we add a runtime call with `strict_metaschema=True`, I see a use-case where cloud-init may apply strict schema verification during runtime for both meta schema and module config schema. Maybe cloud-init would warn on invalid meta schema defintions (because that only affects docs) and error on invalid "config" schema.  Also our existing unit tests in this branch pass either way with the empty config.

Given that is is a "future" discussion, and we don't have any runtime call-sites passing `strict_metadata=True`, it's definitely not something that needs resolution in this PR.

Reply:
@blackboxsw I see what you are getting at. Mutual exclusion is not a requirement. Nothing is stopping us from doing strict metadata validation at runtime. Once this lands and we feel comfortable with how it's working that will be a trivial switch to flip.

For now I've refactored a little for readability and added a test that asserts that a bad schema throws the correct exception. I think the change I just made makes it more obvious _what_ failed when a test fails.